### PR TITLE
feat(relayer): implement graceful shutdown

### PR DESCRIPTION
**Pull request by SuperFluffy**
## Summary
Adds graceful shutdown to relayer.

## Background
Because relayer did not previously respond to SIGTERM, it would eventually just get killed off (for example by kubernetes after 30s). This could potentially lead to situations where relayer would try to submit blocks to Celestia right before shutdown, potentially leaving its state tracking file in a bad state.

## Changes
- Add a SIGTERM listener to sequener-relayer, initiating a shutdown of all its services.
- The relayer task proper will immediately exit its event-loop and wait until its last submission is complete.
- Sequencer-Relayer itself waits 25 seconds for its relayer service to finish, hard-killing it otherwise
- The API task is issues an explicit shutdown signal last and has 4s to finish, before being killed.

## Testing
This should be tested end to end by sending sequencer-relayer SIGTERM

## Related Issues
Closes https://github.com/astriaorg/astria/issues/357

### CODE REVIEW

**Comment by SuperFluffy** [comment]
> have we done any testing with this?

Other than running and kiling it, how would you test this?

File: crates/astria-sequencer-relayer/src/api.rs [hunk 1]
```diff
@@ -39,7 +39,11 @@ impl FromRef<AppState> for RelayerState {
     }
 }
 
-pub(crate) fn start(socket_addr: SocketAddr, relayer_state: RelayerState) -> ApiServer {
+pub(crate) fn start(
+    socket_addr: SocketAddr,
+    relayer_state: RelayerState,
+    // shutdown: oneshot::Receiver<()>,
```
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-sequencer-relayer/src/api.rs:45]: oversight, thought I removed everything

File: crates/astria-sequencer-relayer/src/relayer/mod.rs [hunk 2]
```diff
@@ -192,11 +201,17 @@ impl Relayer {
             );
         };
 
-        submitter_task.abort();
-        if let Err(error) = crate::utils::flatten(submitter_task.await) {
-            error!(%error, "Celestia blob submission task failed");
+        match &reason {
+            Ok(reason) => info!(reason, "shutting down"),
+            Err(reason) => error!(%reason, "shutting down"),
```
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-sequencer-relayer/src/relayer/mod.rs:206]: Yeah, that makes more sense

File: crates/astria-sequencer-relayer/src/relayer/write/mod.rs [hunk 3]
```diff
@@ -235,8 +242,14 @@ impl BlobSubmitter {
             .await
             .wrap_err("failed fetching latest celestia height after many retries")?;
 
-        loop {
+        let reason = loop {
             select!(
+                biased;
```
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-sequencer-relayer/src/relayer/write/mod.rs:247]: Yeah, that orders makes sense to me. Updated

**Review by joroshiba — APPROVED** [review]

**Comment by joroshiba** [comment]
I manually tested this branch locally, confirming shutdown works and submission still working

**Pull request merged**